clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobStreamerImpl.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.worker;
+
+import io.camunda.zeebe.client.api.ZeebeFuture;
+import io.camunda.zeebe.client.api.command.FinalCommandStep;
+import io.camunda.zeebe.client.api.command.StreamJobsCommandStep1.StreamJobsCommandStep3;
+import io.camunda.zeebe.client.api.response.ActivatedJob;
+import io.camunda.zeebe.client.api.response.StreamJobsResponse;
+import io.camunda.zeebe.client.api.worker.BackoffSupplier;
+import io.camunda.zeebe.client.api.worker.JobClient;
+import io.camunda.zeebe.client.impl.Loggers;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import net.jcip.annotations.GuardedBy;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+
+@ThreadSafe
+final class JobStreamerImpl implements JobStreamer {
+  private static final Logger LOGGER = Loggers.JOB_WORKER_LOGGER;
+
+  private final JobClient jobClient;
+  private final String jobType;
+  private final String workerName;
+  private final Duration timeout;
+  private final List<String> fetchVariables;
+  private final Duration requestTimeout;
+  private final BackoffSupplier backoffSupplier;
+  private final ScheduledExecutorService executor;
+  private final Lock streamLock;
+
+  @GuardedBy("streamLock")
+  private ZeebeFuture<StreamJobsResponse> streamControl;
+
+  @GuardedBy("streamLock")
+  private FinalCommandStep<StreamJobsResponse> command;
+
+  @GuardedBy("streamLock")
+  private boolean isClosed;
+
+  @GuardedBy("streamLock")
+  private long retryDelay;
+
+  public JobStreamerImpl(
+      final JobClient jobClient,
+      final String jobType,
+      final String workerName,
+      final Duration timeout,
+      final List<String> fetchVariables,
+      final Duration requestTimeout,
+      final BackoffSupplier backoffSupplier,
+      final ScheduledExecutorService executor) {
+    this.jobClient = jobClient;
+    this.jobType = jobType;
+    this.workerName = workerName;
+    this.timeout = timeout;
+    this.fetchVariables = fetchVariables;
+    this.requestTimeout = requestTimeout;
+    this.backoffSupplier = backoffSupplier;
+    this.executor = executor;
+
+    streamLock = new ReentrantLock();
+  }
+
+  @Override
+  public void close() {
+    try {
+      streamLock.lockInterruptibly();
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
❓Why do we do another interrupt  since the current thread is already interrupted?

Reply:
When you catch a `InterruptedException`, the `interrupt` flag on the thread has already been toggled off. This means `Thread.current.isInterrupted()` will return false from now on. We turn it on again so that callers (which may be blocked interruptibly as well) can be notified.

Reply:
Ahh, I wasn't aware of that logic. Now that makes sense! 👍